### PR TITLE
08817 iss pces replay

### LINKED ISSUE #8817: Don't ingest state signatures when restarting after ISS
When a node restarts following an ISS, it will get another ISS as it replays the signatures in the PCES. If nodes are configured to kill themselves when they detect an ISS, this will cause them to enter a boot-loop and they will be unable to rejoin the network without human intervention.

To compensate for this, we should ignore signature transactions in the PCES when rebooting following an ISS.

### PR DESCRIPTION
**Pull request by cody-littley**
Closes #8817

### CODE REVIEW

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/Scratchpad.java [hunk 1]
```diff
@@ -214,174 +103,32 @@ public <V extends SelfSerializable> V set(@NonNull final K key, @Nullable final
...
-    /**
-     * Generate the path to the next scratchpad file.
-     *
-     * @return the path to the next scratchpad file
-     */
-    @NonNull
-    private Path generateNextFilePath() {
-        return scratchpadDirectory.resolve((nextScratchpadIndex++) + SCRATCHPAD_FILE_EXTENSION);
-    }
+    // TODO test
```
Comment (edward-swirldslabs) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/Scratchpad.java:108]: todo comment

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/iss/IssHandler.java [hunk 2]
```diff
@@ -139,6 +147,28 @@ public void stateHashValidityObserver(
         }
     }
 
+    // TODO test
```
Comment (edward-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/iss/IssHandler.java:150]: todo comment

**Review by edward-swirldslabs — APPROVED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/Scratchpad.java (hunk 1 above)
Comment (cody-littley) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/Scratchpad.java:108]: Test added, TODO removed.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/iss/IssHandler.java (hunk 2 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/iss/IssHandler.java:150]: Coverage added, TODO removed

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+     * the number increasing by one each time a file is written. If multiple files are present, the file with the
+     * highest number is the most recent and is always used. Multiple files will only be present if the platform crashes
+     * after writing a file but before it has the opportunity to delete the previous file.
+     */
+    public static final String SCRATCHPAD_DIRECTORY_NAME = "scratchpad";
+
+    /**
+     * The file extension used for scratchpad files.
+     */
+    public static final String SCRATCHPAD_FILE_EXTENSION = ".scr";
```
Comment (alittley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java:81]: any reason not to use `.scratch`, for the sake of clarity?

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+
+    private final Set<K> fields;
+    private final Map<Integer, K> indexToFieldMap;
```
Comment (alittley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java:84]: Only used in constructor - could optionally be made into a local variable

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+                        {}""",
+                id,
+                table.render());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public <V extends SelfSerializable> V get(@NonNull final K key) {
```
Comment (alittley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java:169]: method should have `@Nullable` annotation

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+        try (final Locked ignored = lock.lock()) {
+            return (V) data.get(key);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public <V extends SelfSerializable> V set(@NonNull final K key, @Nullable final V value) {
```
Comment (alittley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java:179]: `@Nullable`

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/Round.java [hunk 7]
```diff
@@ -42,6 +43,7 @@ public interface Round extends Iterable<ConsensusEvent> {
      * @return an iterator of consensus events
      */
     @Override
+    @NonNull
```
Comment (alittley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/Round.java:46]: Annotated `@NonNull`, but overridden methods are not annotated

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/iss/ConsensusHashManager.java [hunk 8]
```diff
@@ -103,14 +113,12 @@ public class ConsensusHashManager {
...
-     * @param addressBook            the address book for the network
-     * @param currentEpochHash       the current epoch hash
-     * @param currentSoftwareVersion the current software version
-     * @param ignoredRound           a round that should not be validated. Set to {@link #DO_NOT_IGNORE_ROUNDS} if all
-     *                               rounds should be validated.
+     * @param time                         provides the current wall clock time
+     * @param dispatchBuilder              responsible for building dispatchers
+     * @param addressBook                  the address book for the network
+     * @param currentEpochHash             the current epoch hash
+     * @param ignorePreconsensusSignatures If true, ignore signatures from the preconsensus event stream, otherwise
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/iss/ConsensusHashManager.java:120]: you deleted the doc for `currentSoftwareVersion` and `ignoredRound`

**Review by alittley — APPROVED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java (hunk 4 above)
Comment (cody-littley) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java:84]: It's used by a function, but that function is only called by the constructor, so I agree it shouldn't be a class member variable. Refactored.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java (hunk 5 above)
Comment (cody-littley) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java:169]: added

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void logContents() {
+        final TextTable table = new TextTable().setBordersEnabled(false);
+
+        for (final K field : fields) {
+            final SelfSerializable value = data.get(field);
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java:148]: correct

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java (hunk 6 above)
Comment (cody-littley) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/internal/StandardScratchpad.java:179]: added

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/Round.java (hunk 7 above)
Comment (cody-littley) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/Round.java:46]: fixed

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/iss/ConsensusHashManager.java (hunk 8 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/iss/ConsensusHashManager.java:120]: merge conflict gone sideways, restored the missing param docs

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/iss/IssScratchpad.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+package com.swirlds.platform.state.iss;
+
+import com.swirlds.common.scratchpad.ScratchpadType;
+
+/**
+ * Describes the data held in the ISS scratchpad.
+ */
+public enum IssScratchpad implements ScratchpadType {
+    /**
+     * The round number of the most recently observed ISS, or -1 if this node has never observed an ISS.
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/iss/IssScratchpad.java:26]: You are correct, should be `The round number of the most recently observed ISS, or null if this node has never observed an ISS.`.

Fixed.

**Review by alittley — APPROVED** [review]

**Pull request merged**